fil_actor_interface/src/builtin/market/mod.rs

```diff
@@ -129,23 +180,71 @@ pub enum BalanceTable<'a, BS> {
     UnusedBalanceTable(PhantomData<&'a BS>),
 }
 
-pub enum DealProposals<'a, BS> {
-    UnusedDealProposal(PhantomData<&'a BS>),
+pub enum DealProposals<'bs, BS> {
+    V9(V9DealArray<'bs, BS>),
+    V10(V10DealArray<'bs, BS>),
+    V11(V11DealArray<'bs, BS>),
+    V12(V12DealArray<'bs, BS>),
 }
 
 impl<BS> DealProposals<'_, BS> {
     pub fn for_each(
         &self,
-        _f: impl FnMut(u64, DealProposal) -> anyhow::Result<(), anyhow::Error>,
+        mut f: impl FnMut(u64, DealProposal) -> anyhow::Result<(), anyhow::Error>,
     ) -> anyhow::Result<()>
     where
         BS: Blockstore,
     {
-        unimplemented!()
+        match self {
+            DealProposals::V9(deal_array) => {
+                deal_array.for_each(|key, deal_proposal| {
+                    f(
+                        key,
+                        deal_proposal
+                            .try_into()
+                            .expect("failed to convert V9 deal proposal to builtin version"),
```

Comment:
Well, we moved the error handling one lever higher but it's not high enough. Ideally, the library code should **never** panic, unless there was a huge violation of expectations to our own logic. Can we let the calling code deal with this?

All in all, the goal is to evade crashing the application unless it doesn't make sense for it to continue.